maxim/layers.py

```diff
@@ -1,65 +1,89 @@
 """
 Layers based on https://github.com/google-research/maxim/blob/main/maxim/models/maxim.py
+and reworked to cope with variable image dimensions
 """
 
-import einops
 import tensorflow as tf
 from tensorflow.experimental import numpy as tnp
-from tensorflow.keras import backend as K
 from tensorflow.keras import layers
 
 
 @tf.keras.utils.register_keras_serializable("maxim")
-class BlockImages(layers.Layer):
+class TFBlockImages(layers.Layer):
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
 
-    def call(self, x, patch_size):
-        bs, h, w, num_channels = (
-            K.int_shape(x)[0],
-            K.int_shape(x)[1],
-            K.int_shape(x)[2],
-            K.int_shape(x)[3],
-        )
+    def call(self, image, patch_size):
+        bs, h, w, num_channels = (tf.shape(image)[0], tf.shape(image)[1], tf.shape(image)[2], tf.shape(image)[3])
+        ph, pw = patch_size
+        gh = h // ph
+        gw = w // pw
+        pad = [[0, 0], [0, 0]]
+        patches = tf.space_to_batch_nd(image, [ph, pw], pad)
+        patches = tf.split(patches, ph * pw, axis=0)
+        patches = tf.stack(patches, 3)  # (bs, h/p, h/p, p*p, 3)
+        patches_dim = tf.shape(patches)
+        patches = tf.reshape(patches, [patches_dim[0], patches_dim[1], patches_dim[2], -1])
+        patches = tf.reshape(patches, (patches_dim[0], patches_dim[1] * patches_dim[2], ph * pw, num_channels))
+        return [patches, gh, gw]
```

Comment:
I'm honestly not sure why we are getting rid of einops. This is significantly more lines of code and also more complex to read.

Reply:
Indeed, using einops would be in hand. But please, consider this code snippet:
```
img = tf.random.uniform((1, 4, 4, 1))
block_img = einops.rearrange(img, 'b (gh fh) (gw fw) c -> b (gh gw) (fh fw) c', fh=2, fw=2) # this should work fine
block_img_with_tensors_as_arguments = einops.rearrange(img, 'b (gh fh) (gw fw) c -> b (gh gw) (fh fw) c', fh=tf.constant([2]), fw=tf.constant([2])) # this breaks.
```
The problem with einops is that it expects int literals as argument to the symbols used in the pattern string. I could not make it work using tensors as the example above. At some stages of the model ([here](https://github.com/sayakpaul/maxim-tf/blob/6364fd25f39b4dcc73565db56d31c2d762376f72/maxim/blocks/misc_gating.py#L126), [here](https://github.com/sayakpaul/maxim-tf/blob/6364fd25f39b4dcc73565db56d31c2d762376f72/maxim/blocks/misc_gating.py#L136), [here](https://github.com/sayakpaul/maxim-tf/blob/6364fd25f39b4dcc73565db56d31c2d762376f72/maxim/blocks/misc_gating.py#L121)), the split is computed in online fashion, thus relying on tensors (for the case where the img size is None). Thus it was necessary to rewrite using tensorflow.

Reply:
But it wasn't a problem with the current version of the code. What changed?

Reply:
The current version of the code informs the image dimension beforehand, thus when you do:
```
n, h, w, num_channels = (
            K.int_shape(x)[0],
            K.int_shape(x)[1],
            K.int_shape(x)[2],
            K.int_shape(x)[3],
        )
```
you have the integer literals we need for the einops operations. However, In case when we feed (None, None, 3) as input, `h` and `w` cannot be used for computing direct literals for the einops operations, and they have to be represented as tensorflow placeholders (None tensor).